src/ci/docker/host-x86_64/x86_64-uefi/uefi_qemu_test/run.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+
+from pathlib import Path
+
+
+def run(*cmd, env=None, capture=False):
+    """Print and run a command, optionally capturing the output."""
+    cmd = [str(p) for p in cmd]
+    print(' '.join(cmd))
+    return subprocess.run(cmd,
+                          capture_output=capture,
+                          check=True,
+                          env=env,
+                          text=True)
+
+
+def build_and_run(tmp_dir):
+    host_artifacts = Path('/checkout/obj/build/x86_64-unknown-linux-gnu')
+    stage0 = host_artifacts / 'stage0/bin'
+    stage2 = host_artifacts / 'stage2/bin'
+
+    env = dict(os.environ)
+    env['PATH'] = '{}:{}:{}'.format(stage2, stage0, env['PATH'])
+    env['RUSTFLAGS'] = '-C linker=lld-link'
+
+    # Copy the test create into `tmp_dir`.
+    test_crate = Path(tmp_dir) / 'uefi_qemu_test'
+    run('cp', '-r', '/uefi_qemu_test', test_crate)
+
+    # Build the UEFI executable.
+    target = 'x86_64-unknown-uefi'
+    run('cargo',
+        'build',
+        '--manifest-path',
+        test_crate / 'Cargo.toml',
+        '--target',
+        target,
+        env=env)
+
+    # Create a mock EFI System Partition in a subdirectory.
+    esp = test_crate / 'esp'
+    boot = esp / 'efi/boot'
+    os.makedirs(boot, exist_ok=True)
+
+    # Copy the executable into the ESP.
+    src_exe_path = test_crate / 'target' / target / 'debug/uefi_qemu_test.efi'
+    shutil.copy(src_exe_path, boot / 'bootx64.efi')
```

Comment:
You use `run('cp', ...)` before, and now `shutil.copy()`. For consistency we could replace the first with `shutil.copytree()`? But maybe this is fine. I am not sure what the rules for CI and external dependencies is. coreutils should be fine?

Reply:
I would prefer to be consistent with copytree. But in general as long as it only uses packages from apt repos it's not a big deal, please ask me before downloading random binaries though.

Reply:
Updated to use `copytree`.

> please ask me before downloading random binaries though

I think we should be good there, we're only doing apt downloads.